lib/trino-parquet/src/main/java/io/trino/parquet/reader/ChunkedInputStream.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.parquet.reader;
+
+import com.google.common.collect.Iterators;
+import io.airlift.slice.BasicSliceInput;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+import io.trino.memory.context.AggregatedMemoryContext;
+import io.trino.memory.context.LocalMemoryContext;
+import io.trino.parquet.ChunkReader;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Collection;
+import java.util.Iterator;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkPositionIndexes;
+import static com.google.common.io.ByteStreams.readFully;
+import static java.util.Objects.requireNonNull;
+
+public class ChunkedInputStream
+        extends InputStream
+{
+    private final LocalMemoryContext dataReaderMemoryUsage;
+    private final Iterator<BasicSliceInput> chunks;
+    private BasicSliceInput current;
+
+    public ChunkedInputStream(Collection<ChunkReader> chunks, AggregatedMemoryContext memoryContext)
+    {
+        this.dataReaderMemoryUsage = requireNonNull(memoryContext, "memoryContext is null")
+                .newLocalMemoryContext(ChunkedInputStream.class.getSimpleName());
+        requireNonNull(chunks, "chunks is null");
+        checkArgument(!chunks.isEmpty(), "At least one chunk is expected but got none");
+        this.chunks = Iterators.transform(chunks.iterator(), this::readChunk);
+        current = this.chunks.next();
+    }
+
+    public Slice getSlice(int length)
+    {
+        if (length == 0) {
+            return Slices.EMPTY_SLICE;
+        }
+        while (!current.isReadable()) {
+            checkArgument(chunks.hasNext(), "Requested %s bytes but 0 was available", length);
+            current = chunks.next();
+        }
+        if (current.available() >= length) {
+            return current.readSlice(length);
+        }
+        // requested length crosses the slice boundary
+        byte[] bytes = new byte[length];
+        try {
+            readFully(this, bytes, 0, bytes.length);
+        }
+        catch (IOException e) {
+            throw new RuntimeException("Failed not read " + length + " bytes", e);
```

Comment:
I fixed the typo.
For chunkKey. The source exception should have all the required data. It doesn't make sense to push chunkKey here just to include it in the message.
I will extend `AbstractParquetDataSource` to put this info in the exception

Reply:
The `extendExceptionInfo` change looks overkill for this, I would definitely avoid that change even at the cost of not getting the extra info in the exception.
`ChunkReader` provides a `getDiskOffset` method which might be useful to log instead of ChunkKey. If you'd prefer to avoid that then I suggest we stick with the code before `extendExceptionInfo`.

Reply:
`extendExceptionInfo` droped